Convert project to TypeScript

bb opened the pull request:
This is a rough conversion of the project to TypeScript. With rough I mean some usage of `any` here and there, some ad-hoc type definitions and many eslint rules disabled. This fixes #118 and #121. It also sneaks in the fix for Contributing Docs (same as meilisearch/meilisearch-js#687).

Everything should work at least as good as before. It still uses the default export (#122) because the way it is written now, it can be used in another TypeScript project.

bb commented:
The failing integration-test is because https://github.com/meilisearch/meilisearch-js/pull/688 is not yet released.

If you check out meilisearch-js locally and use it by running `yarn link` in meilisearch-js and `yarn link meilisearch` in instant-meilisearch, it will work.

bb commented:
bors try

curquiza commented:
bors try

curquiza commented:
Hello @bb!
Thanks a lot for this work! This is a huge start to move this project to typescript!!
Because it might need some improvements, we created a `typescript-dev` branch and we will merge your work into this branch as a first step 🙂
@bidoubiwa and I will review your PR and will continue your work on the `typescript-dev` branch! When all will be ready, we will merge `typescript-branch` into `master`.
Thanks again for everything!!! 🚀

curquiza reviewed (commented):
Thanks for this huge work! We are waiting for the @bidoubiwa's review.

A small change maybe if you have time @bb: could you remove the changes in the CONTRIBUTING.md? It's not related to the typescript work.

bidoubiwa commented on src/index.ts:
```diff
@@ -1,18 +1,101 @@
+      index: string
+      hitsPerPage: number
+      facets: object | undefined
+      exhaustiveFacetsCount?: boolean
+      processingTimeMs: number
+      exhaustiveNbHits: boolean
+    }>
+  }
+
+  transformToMeiliSearchParams: (
+    params: Params
+  ) => {
+    q: any
+    facetsDistribution: any
+    facetFilters: any
+    attributesToHighlight: string[]
+    attributesToCrop: string[] | undefined
+    filters: any
+    limit: number
+  }
+  parseHits: (meiliSearchHits: any, params: Params) => any
+  paginationParams: (
+    hitsLength: number,
+    params: Params
+  ) => { nbPages: number; page: number | undefined } | undefined
+
+  search: (
+    requests: Requests
+  ) => Promise<{
+    results: Array<{
+      hits: any
+      index: string
+      hitsPerPage: number
+      facets: object | undefined
+      exhaustiveFacetsCount?: boolean | undefined
+      processingTimeMs: number
+      exhaustiveNbHits: boolean
+    }>
+  }>
+}
```
Should these types not be exported as well?

bb replied:
I didn't need them, but they might be useful to someone. Should I add the exports or should we wait until they're requested?

curquiza replied:
No need to export them for this first PR 🙂 Let's keep it like this for now.

bidoubiwa commented on src/format.ts:
```diff
@@ -1,6 +1,11 @@
+import { isString } from './utils'
 
-function replaceHighlightTags(value, highlightPreTag, highlightPostTag) {
+function replaceHighlightTags(
+  value: string,
+  highlightPreTag?: string,
+  highlightPostTag?: string
+) {
```
The return type is missing or is it infered based on the returned function?

bb replied:
Yes, it's inferred. The return statement explicitly states `.toString()`.

However, there are other things which might not be perfect yet: Is the input param really always string as specified? Currently it looks like, but why would it need to check isString() in line 15 then?

bidoubiwa commented on src/format.ts:
```diff
@@ -15,30 +20,30 @@ function replaceHighlightTags(value, highlightPreTag, highlightPostTag) {
   highlightPreTag,
   highlightPostTag,
-}) {
+}: { formattedHit: T } & Params) {
```
Why is it both `{ formattedHit: T }` and `Params` as return types?

bb replied:
Because formattedHit is not defined in Params.

If you remove it, line 24 will be marked as error:
> Property 'formattedHit' does not exist on type '{ page?: number | undefined; hitsPerPage?: number | undefined; highlightPreTag?: string | undefined; highlightPostTag?: string | undefined; snippetEllipsisText?: string | undefined; attributesToSnippet?: string[] | undefined; query: string; facets: string[]; } & SearchParams<...>'.ts(2339)

However, it would probably be a good idea to refactor this into a separate named type. One of the rough edges I mentioned in the initial description 😉

bidoubiwa reviewed (commented):
Thanks a lot for this work! I see you took the configuration of meilisearch-js. Can I ask you if you think some things should be improved in the typescript configuration? In order to know on what to work. 

I asked you some questions I hope they are relevant

bb commented:
> A small change maybe if you have time @bb: could you remove the changes in the CONTRIBUTING.md? It's not related to the typescript work.

Good point. Sorry for mixing this stuff. I now committed the contributing file from current master on top, so all should be fine again.

bb commented:
> Thanks a lot for this work! I see you took the configuration of meilisearch-js. Can I ask you if you think some things should be improved in the typescript configuration? In order to know on what to work.

I thought it would be a good idea to have both projects not diverge too much, so I copied some configs from there. I think the config itself is fine.

Things to improve in general:
* there's one Todo comment left (index.ts line 14)
* there are quite a few `any` type annotations, esp. in `InstantMeiliSearchInstance`
  * an especially nasty one is `(result[key] as any)` in format.ts. I didn't find a good solution within a few minutes of googling, so I left it as is.
* there are a few `foo as Bar` casts, which are probably also a code smell. 
  * e.g. the `Object.keys(object) as K[])` in utils.ts line 9
* It would probably be nice to enable generic Parametrization of the Index's structure (which currently uses `    meiliSearchResponse: SearchResponse<any, any>`)

As you can see, it's all not critical to fix but also not a great clean state yet.

Does this overview help as a starter?

curquiza reviewed (approved):
I also realize we have to find a way to activate the "auto-reload/rebuild"  for the playgrounds because of this typescript integration, otherwise, it will be tedious to develop with them.

Thanks for your work @bb, there is only a small git conflict (I made the `typescript-dev` breach up-to-date with `master`).
Afterward, I'm going to merge your PR. @bidoubiwa will continue your work. She's on vacations right now and has emergency tasks on her todo list before doing this one, but be sure she's looking forward to working on it 😉
And thanks for taking the time to answer our questions and to suggest improvements! 🚀

bidoubiwa reviewed (approved):
Thanks for the amazing work

bb commented:
I just fixed the merge conflict in package.json and modified the new vanilla-js taks to include the build step directly in the Github UI, but it looks like this was not enough. Will check locally now.

bb commented:
I resorted to an entry in the ignore file to make the linter pass with the TypeScript base settings and the new vanilla-js parts. Not perfect, but better than failing checks.

curquiza commented:
Merging then!! Thanks a lot @bb for all your involvement!! 🎉
